wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+the metadata for all of its callees. There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the AST of the function, building a directed graph along the way.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+Note: Each vertex in this graph corresponds either to the statement "The control-flow at a specific program point must be uniform" or to "The value of a specific expression must be uniform" or to "A specific variable must always hold a uniform value" or is one of the special vertices True and False.
+      Each (directed) edge corresponds to an implication.
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some vertices called "True", "False", "CF_start", "CF_return" and if the function is non-void a vertex called "Value_return"
+* Create one vertex for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding vertices and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body treated as a statement (see next section):
+    * If it has a behavior which is neither Nothing nor Discard, then reject the program
+    * Else remember it for later it.
+* Look at which vertices are reachable from "True"
+    * If this set includes the vertex "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "Value_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. Also note that the information that we need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
```

Comment:
I'd appreciate a statement describing the meaning of the set. Something like "this is the set of the possible ways that control flow can exit the statement."

Reply:
It is exactly that. I'll add that sentence.

Reply:
At this point, the meaning of "Nothing" seems confusing to me. If each statement is associated with a set of behaviors, wouldn't the empty set be the same as "Nothing"?

Reply:
Nothing means that control can flow to the next statement.
I wanted to call it Fallthrough, but it conflicted with the actual Fallthrough statement. Maybe "Default" or "NextStatement" might be clearer?

The empty set would mean that control flow is trapped forever in the statement.